프로젝트 매니저 [STEP 2] 써니쿠키

sunny-maeng opened the pull request:
안녕하세요 스티븐!(@stevenkim18) Step2 PR드립니다 🙇🏻‍♀️
MVVM 구현에 초점을 맞췄는데, 이렇게 하는 게 맞는지, 계속 한 쪽 머리에 물음표를 달고 프로젝트를 진행했던 것 같습니다ㅎㅎ.. 
쓴조언도 환영합니다 ㅎㅎ

## 📝 STEP 진행 중 경험하고 배운 것
- MVVM 디자인패턴
	 - [X] MVVM디자인 패턴을 공부하고 적용해보았습니다.
	 - [X] Closure를 이용해 ViewModel과 View를 바인딩했습니다.
	 - [X] ViewModel이 변경되면 didSet으로 불리는 binding Closure를 통해 View가 자동으로 업데이트 되도록 구현했습니다.
- ViewModel의 UnitTest
	- [X] MVVM의 장점으로 ViewController에서 ViewModel로 분리된 로직들의 UnitTest가 가능하다는 장점을 살려, ViewModel의 UnitTest를 진행했습니다.
- modal의 formSheet, Alert의 PopoverStyle 사용
	 - [X] 등록/수정 뷰를 전체화면 중 중앙에 작게 보여주기위해 modal의 formSheet로 화면을 보여줍니다.
	 - [X] Cell의 이동 옵션을 메뉴로 보여주기 위해 Alert의 PopoverStyle을 사용했습니다.
- UIGestureRecognizer
     - [X] Cell을 길게 터치했을 때 액션을 추가하기 위해 Cell에 `UILongPressGestureRecognizer`로 Gesture를 추가했습니다. 

## ⚒️ 코드 구현내용 
### 1️⃣ Process 열거형 / Project 구조체
- Process (M)
	- 프로젝트의 진행과정을 나누어 TODO, DOING, DONE 세 과정으로 열거합니다.
	- 세 과정과 연관되어 배열로 관리 될 데이터에서 사용 할 `index`를 제공합니다.
- Project (M)
	- 하나의 프로젝트를 나타내는 모델입니다
	- `diffableDataSource`에서 데이터의 고유성을 위해 `UUID`를 추가로 갖습니다.

### 2️⃣ MainViewModel - MainViewController
세 개의 프로젝트 과정(TODO/ DOING/ DONE)을 리스트 세 개로 보여주고 컨트롤 합니다.
프로젝트 등록화면으로 이동하거나, 셀 선택 시 등록된 프로젝트를 자세히 보여주는 화면으로 이동할 수 있습니다
셀을 길게 누르면 셀을 다른 리스트로 옮길 수 있는 옵션메뉴를 띄어줍니다.
- MainViewController (V)
	- lists 세 개를 그리고 layout 합니다.
	- lists 세개와 DataSource 세 개는 배열로 관리됩니다.
	- MainViewModel의 `datas`가 변경 될 때마다 호출 될 `updateDatas`클로저를 바인딩해, 데이터 변경시마다 `dataSource`에 새로운 `snapshot`을 찍습니다.
- MainViewModel (VM)
	- 세 개의 리스트에 담길 데이터가 2차원 배열인 `datas`를 갖습니다.
	- 데이터를 추가/ 삭제/ 수정/ 이동/ 읽기를 담당하는 메서드들을 갖고, 이 메서드들을 사용해 `datas`배열 아이템들을 다룰 수 있습니다.
	- `datas`가 변할 때마다 `updateDatas`클로저를 호출합니다.
	- 그 외, 뷰에 보여줄 `ListTitles`배열, `datasCount`를 갖습니다.

### 3️⃣ ListCellViewModel - ListCell
리스트의 각 Cell을 컨트롤합니다.
- ListCell (V)
	- 제목 / 목표날짜 / 상세내용요약 을 나타낼 View를 그리고 layout합니다.
	- `ListCellViewModel`의 제목데이터, 날짜데이터, 내용데이터가 변경될 때마다 호출 될 클로저들을 바인딩해, 데이터 변경시마다 `label`로 보여주는 Text를 변경합니다.
	- 길게 터치 시, 액션을 추가하기 위해 UILongPressGestureRecognizer가 추가되어있습니다.
- ListCellViewModel (VM)
	- `title` / `date` / `description` 을 String 타입으로 갖습니다.
	- 목표날짜를 오늘과 비교해 기한이 지났는지 확인합니다.(`isMissDeadLine`)
	- cell의 내용을 구성합니다(`setupCell`)

### 4️⃣ EditingViewModel - EditingViewController
리스트에 새로운 프로젝트를 추가하는 뷰/ 추가 되어있던 프로젝트를 자세히 보고, 수정할 수 있는 뷰를 컨트롤합니다.
- EditingViewController (V)
	- Editable모드/ ReadOnly모드가 있습니다.
	- 제목, 상세내용, 목표날짜를 작성(수정)할 수 있는 View를 그리고 layout합니다.
	- `EditingViewModel`의 제목데이터, 날짜데이터, 내용데이터가 변경될 때 호출되는 클로저들을 바인딩해, 데이터를 보여줍니다.
	- Editable모드시 취소, 등록 버튼을 보여주고, ReadOnly모드일 때는 수정, 등록 버튼을 보여줍니다. 
- EditingViewModel (V)
	- Editable모드/ ReadOnly모드구분하고 `title` / `date` / `description` 을 String 타입으로 갖습니다.
	- 새로운 프로젝트라면 새로운 데이터를 추가하고, 기존 프로젝트라면 해당 데이터를 수정합니다. 이 때, `MainViewModel`의 `datas`를 변경해줍니다.

---

## 💭 고민한 부분
### 👉 리스트 그리기 -`UICollectionView`/ `UITableView`
View를 먼저 그릴 때, 어떤 방법으로 리스트를 그릴 지 고민했습니다
- 1. `UICollectionViewCompositionalLayout.list`와 `Section` 3개
	- CollectionView 에서 지원해주는 List형태를 이용해 한개의 CollectionView에서 Section을 .todo/ .doing / .done 으로 구분해 리스트들을 그리려 했습니다.
	- Section으로 구분 시 각 Section 별 리스트들을 수평으로 그리지 못하고, 수직으로만 쌓아 그릴 수 있어서 적용하지 못했습니다. 
- 2. `UICollectionView` 를 `compositionalLayout`으로 List형태 만들기
	- 하나의 `CollectionView`안에 `Section`으로 구분해 수평으로 쌓이는 리스트를 만들기 위해 `UICollectionView`를 `compositionalLayout`으로 옆으로 쌓이는 List형태로 만드는 방법도 고민했었습니다.
	- 이렇게 만들 시, Cell SwipeAction을 지원해주지 않기 때문에 별도로 구현해줘야하기 떄문에 보류했습니다.
- 3. `UITableView`
	- Section으로 구분이 안되니 같은 리스트를 똑같이 세번씩 그려줘야 하는데, `UICollectionViewCompositionalLayout.list`를 세 개 그릴지, `UITableView`를 세 개 그릴지 고민했습니다.
	- `UICollectionViewCompositionalLayout.list`는 Cell Custom이 어려운 단점이 있고, 프로젝트 요구사항에도 cell사이에 간격이 있는것을 보고 Custom이 필요한 것 같아 Cell Custom이 더 쉬운 `UITableView`를 선택했습니다.

### 👉 세 개의 List, DataSource, Snapshot관리
한 화면에 프로젝트 절차인 TODO/ DOING/ DONE에 각각 `UITableView`를 사용해 총 3개 그리다보니 `DataSource`도 세 개, `SanpShot`도 3개로 관리해야했습니다. 

처음엔 각각 세개씩 아홉개의 프로퍼티를 만들고 관리했었는데, 스티븐이 중간점검에서 배열로 관리하고, TODO/ DOING/ DONE 을 열거형으로 만들어놓았던 `Process`타입에서 `index`를 관리할 수 있다는 아이디어를 채택해 **배열관리**로 수정했습니다.

`DataSource`는 `SnapShot`을 포함한 하나의 타입으로 만들었습니다. 최종적으론 `DataSource`와 `List`를 배열로 관리하고 `Process` 열거형 타입에 `index`연산프로퍼티를 구현해 배열들과 Process를 연결했습니다.

### 👉 MVVM의 장점은 무엇인가......UnitTest ?
MVVM 디자인패턴을 처음 공부해보고 접목해보면서 장점이 무엇인가에 대한 생각을 많이 했습니다. 

이곳 저곳에 물어보고 다니면서, MVVM의 장점이 ViewController에서 로직이 분리되어 나왔기 때문에 유닛테스트를 더 쉽게 할수있다는 것과, 방대해지는 ViewController의 일을 분담시킬 수 있다는 점이 있었습니다.

MVVM의 장점을 살려보고자 프로젝트에서 최대한 많은 로직을 ViewModel로 분리하고 싶었으나 **UIKit**을 import하지않고 분리해낼 수 있는 부분이 이번 프로젝트에선 많지 않았던 것 같다고 생각했습니다.. 

확실히 **UnitTest**는 쉽게 진행할 수 있었던 것 같습니다. ViewController에 있는 로직을 꼭 UnitTest해야하는 경우라면, VeiwModel로 분리해내면 좀 더 쉽게 테스트 할 수 있다는 것을 새로 배울 수 있었습니다.

---

## 🧐 조언을 얻고 싶은 부분
### 👉 MVVM관련
1. 뷰모델이 다른 뷰모델을 프로퍼티로 가지고있어도 되는지 궁금합니다.
예를들면 등록된 프로젝트를 수정하려고 할 때, EditingViewModel은 MainViewModel의 어떤 데이터(Project)인지 전달받야하고, 수정이끝나면 다시 그 데이터가 이렇게 바뀌었다고 전달해줘야하는 상황이라 지금 같은경우엔,  `EditingViewModel` 에서 `var editTargetModel: MainViewModel`
을 프로퍼티로 갖고있어 이니셜라이저에서 MainViewModel을 넘겨줘야하는 로직입니다..
이렇게 구현하는게 맞는 방향인지 조언을 얻고싶습니다 .. 🤔

2. 다른 MainViewController 에게 EditViewModel을 매개변수로 넘겨줘도 되는지 궁금합니다. 
롱터치액션 핸들링에서, cell에 LongPressGestureRecognizer를 추가하고 delegate로 MainViewController로 핸들링을 넘겨줄 때, 어떤 프로젝트에 대한 정보인지 넘겨주기위해 cellViewModel을 매개변수에 담아서 넘겨주는데 이 방법이 괜찮은 방법인지 조언을 얻고 싶습니다.. 🤔

sunny-maeng committed:
feat: 코드 UI구현을 위한 사전 준비

- 스토리보드삭제
- navigationController embedded ViewController 연결

sunny-maeng committed:
feat: Model 타입 구현 - Project 구조체

sunny-maeng committed:
design: ListView타입에, tableView구현 및 ProjectCell 구현

- UILabel, UIStackView extension으로 속성을 부여하는 편의이니셜라이저 생성

sunny-maeng committed:
design: ListView구현 - 타이틀, 구분선 추가 후 stack으로 layout구성

sunny-maeng committed:
feat: tableView의 DataSource, SnapShot 생성 및 구성

sunny-maeng committed:
feat: MainView에 toDo / doing / done ListView 추가 후, DataSource, snapShot 생성 및 구성

sunny-maeng committed:
feat: NavigationBar 구성 - Title

sunny-maeng committed:
design: 레이아웃 간격 조정 및 배경색 지정

sunny-maeng committed:
design: 원형으로 커스텀한 CircleLabel 구현 및 CountLabel에 적용

sunny-maeng committed:
refactor: 네이밍 수정 및 줄바꿈 컨벤션 통일

- 네이밍수정: ProjectCell ➡️ ListCell

sunny-maeng committed:
refactor: ListCellViewModel 구현 및 적용

sunny-maeng committed:
feat: MainViewModel 클래스 타입 생성, ListTitle 에 뷰모델 데이터 적용

sunny-maeng committed:
refactor: 3개의 ListView dataSources, snapShots들 배열로  관리하는 로직으로 수정

- 배열관리: Process 열거형에서 Index프로퍼티 생성

sunny-maeng committed:
feat: selected 되는동안 Cell 테두리 blue Color 적용

sunny-maeng committed:
feat: +버튼 터치 시, 등록화면이동 로직 구현: RegisterViewController타입구현, FormSheet 형태, modal 방식

sunny-maeng committed:
design: 등록화면에 titleField, dataPicker 생성 및 레이아웃

- UITextField Extension구현: 편의이니셜라이저로 PlaceHolder지정, addPadding메서드, addShadow메서드

sunny-maeng committed:
design: EditingViewController에 navigationBar, descriptionTextView구현 및 레이아웃

- refactoring: 타입명 수정 RegisterViewController ➡️ EditingViewController
- UITextView Extension구현: addShadow()메서드, 속성 지정하는 편의이니셜라이저
- todo/ doing / done 중 어느 카테고리의 편집인지 알기위해 process 프로퍼티 구현, 이니셜라이징에서 설정가능

sunny-maeng committed:
feat: 프로젝트 등록 MVVM 적용 - EditingViewModel타입구현, bindingToDoData() 구현

- 뷰컨에있던 process프로퍼티 뷰모델로 이동
- 뷰모델 사이 데이터전달: EditingViewModel에서 MainViewModel프로퍼티로 지정, 이니셜라이징 시 지정하도록 설정
- 뷰모델 - 뷰 사이 데이터전달 : 클로저콜백, 뷰모델에서 updateTodoData클로저생성후 didSet호출,  뷰에서 bindingToDoData()메서드로 클로저 지정

sunny-maeng committed:
feat: DataSource 타입구현 - 초기 스냅샷, 수정시 스냅샷으로 사용할  메서드 구현 및 적용

- 초기스냅샷: applyInitialSnapShot() 메서드
- 데이터소스 수정 후 스냅샷: applySnapshot() 메서드
- MainVC에 적용 - takeInitialSnapShot(), bidingViewModel()

sunny-maeng committed:
feat: MVVM적용 - ProjectCount Label Text 에 적용

sunny-maeng committed:
feat: Date표기 format적용 및 마감날짜확인해 지난날짜 빨간 글씨 처리

- Date Extension - dotFormat() 메서드 구현
- 기한확인 : isMissDeadLine 프로퍼티 구현

sunny-maeng committed:
refactor: 네비게이션바관련 메서드 네이밍 통일, 스타일 통일

sunny-maeng committed:
feat: MVVM적용 - EditingVM, ListCellVM 데이터 바인딩 및, 셀터치 시 EditingView 팝업 띄우기

sunny-maeng committed:
feat: Project 타입에 UUID추가,  프로젝트 수정로직 구현

- 수정화면 넘어갈 때, 기존 mainViewModel에 존재하던 데이터를 넘기도록 수정
- dataSource타입 reload()시 모두 삭제 후 다시 append하는 로직

sunny-maeng committed:
feat: trailing스와이프 삭제기능 구현

sunny-maeng committed:
refactor: todo, doing, done data들 묶어서 datas 배열 관리 및 Cell에 ListCellModel 적용하는 setupCell() 메서드 구현

sunny-maeng committed:
feat: longPressTouch기능 구현 - PopOver 프로젝트 이동 메뉴띄어주고, 메뉴 선택 시 프로젝트 이동

-  Process열거형에 movingOption구현
-  ViewModel에 데이터를 이동하는 moveData() 메서드 구현
- Popover를 띄우는 showPopoverMenu()메서드, 메뉴의 액션들을 생성하는 generateMovingActions()메서드
- ViewModel을 이용해 프로젝트를 이동하는 moveProject() 메서드

sunny-maeng committed:
refactor: class타입 final설정 및 접근제어 설정에 따른 코드수정 및 코드 컨벤션 통일

sunny-maeng committed:
refactor: nameSpace 정리

sunny-maeng committed:
fix: 마감기한이 지난 날짜 빨간글씨 표시 로직 중 오늘 날짜도 빨간색처리되는 버그 수정 - Date 타입 비교 ➡️ Date 날짜단위로 String 변환후 비교

- 원인: Date()타입 비교 시, today(Date())는 초단위 까지 나타나고, 마감기한은 한번 날짜단위로 String처리가 되어서 초단위 존재하지 않음(0초로 나타남)

sunny-maeng committed:
feat: Cell 터치 시 뜨는 Detail 화면 수정 잠금 설정, Bar버튼인 Edit 터치 시 수정 가능하도록 구현

- EditingViewModel에 EditingMode열거형 구현 - editable / readOnly 구분
- 뷰모델의 Mode프로터피 뷰와 바인딩: changeMode() 클로저 사용
- editable / readOnly 에 따른 네비게이션 바 설정 변경: 왼쪽 바버튼이 editable 시 Cancel, readOnly시 Edit

sunny-maeng committed:
test: MainViewModel 유닛테스트 진행

sunny-maeng committed:
test: EditingViewModel 유닛테스트 진행

sunny-maeng committed:
feat: 뷰를 가리지 않도록 키보드 핸들링 - 키보드 노티피케이션 등록 및 키보드 show / hide 시 View 크기 수정

stevenkim18 commented:
써니쿠키!! 1주일 넘는 시간 동안 너무 고생 많으셨습니다!

캠프생활을 아주 열심히 하신 흔적이 곳곳에서 보이네요 ㅎㅎ

코드도 전반적으로 깔끔하고 파일도 정리를 잘하신 것 같습니다.

일단 제가 써니쿠키를 코드를 보면서 전반적으로 아쉬운 부분을 먼저 말씀 드릴게요.

1. 네이밍이 바로바로 와닿지 않는 경우가 있었습니다.
    1. view인데 네이밍을 view인지 알 수 없었다.
    2. 모델 타입이 직관적이지 않다(Project, Process)
2. 클로져를 많이 사용해서 코드의 가독성이 떨어진다.
    1. 클로져 안에 클로져를 사용하셔서 아쉬웠습니다.
    2. 클로져의 생략 표현들을 많이 사용하셨는데 처음 학습할 때는 풀어서 쓰는 것을 저는 권장합니다.
3. Main뷰컨, Edit뷰컨, Cell 모두에 뷰모델이 있어서 오히려 구조가 복잡해진 느낌이었다.
    1. Main 뷰컨에 대한 뷰모델만 있고 필요에 따라서 다른 뷰모델을 만들어도 되지 않을까..

특히 마지막 3번째 이유로 인해서 조언을 얻고 싶은 부분에 2가지 질문이 나오지 않았나 싶습니다.

제가 처음 요청 드렸을 때 MVC패턴으로 작성하시고 리펙토링을 해보시는 걸 추천드렸는데 이런 과정이 고민한 부분에 잘 설명이 되어 있었으면 어땠을까 싶습니다.

이렇게 총평은 마무리하고 세부적으로 코드에 대해서 아쉬운 점들을 댓글로 달아드릴게요!

stevenkim18 commented on ProjectManager/ProjectManager/Extension +/UILabel +.swift:
```diff
@@ -0,0 +1,23 @@
+
+import UIKit
+
+extension UILabel {
+    
+    convenience init(font: UIFont.TextStyle = .body,
+                     textColor: UIColor = .black,
+                     numberOfLines: Int = 0,
+                     textAlignment: NSTextAlignment = .natural) {
+        self.init(frame: .zero)
+        self.font = .preferredFont(forTextStyle: font)
+        self.textColor = textColor
+        self.numberOfLines = numberOfLines
+        self.textAlignment = textAlignment
+        translatesAutoresizingMaskIntoConstraints = false
+    }
+}
```
extension과 편의 생성자의 적극적인 활용 매우 좋습니다! 👍

stevenkim18 commented on ProjectManager/ProjectManager/Extension +/UIView +.swift:
```diff
@@ -0,0 +1,18 @@
+//  UIView +.swift
+//  ProjectManager
+//
+//  Created by 맹선아 on 2023/01/13.
```
?? 본명이신가요? ㅎㅎ

sunny-maeng replied:
으앗 네 본명입니다 수정 완료했습니다😂 [67b99a3](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/67b99a3aa9dc28c5215d3119e2b9d1855ef35d82)

stevenkim18 commented on ProjectManager/ProjectManager/View/EditingView/EditingViewController.swift:
```diff
@@ -0,0 +1,222 @@
+        return textView
+    }()
+    
+    private let stack = UIStackView(axis: .vertical)
```
stackview가 네이밍에 들어가면 더 좋을 것 같아요!

sunny-maeng replied:
모든 View 아이템 들의 네이밍에 View를 붙였 수정했습니다  [67b99a3](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/67b99a3aa9dc28c5215d3119e2b9d1855ef35d82)
- stack ➡️ stackVeiw 
- separatorLine ➡️ separatorLineView

stevenkim18 commented on ProjectManager/ProjectManager/View/MainView/ListView.swift:
```diff
@@ -0,0 +1,106 @@
+    private let headStack = UIStackView(axis: .vertical, alignment: .leading)
+    private let headAndListStack = UIStackView(axis: .vertical,
+                                               spacing: Default.stackSpacing,
+                                               backgroundColor: .systemGroupedBackground)
```
단순한 뷰 인데 StackView를 2개 사용하셨어요!! 이유가 있으신가요???
(여기도 네이밍에 stack view 넣어주세요!)

sunny-maeng replied:
현재 상황에서는 불필요한 View로 확인했고 삭제했습니다. [d22ccb2](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/d22ccb2664835ae7207b86db38867bde21b983c2)

( 지금 설계와 다르게 초반에는 Item 하나를 Section으로 두고 레이아웃을 했었는데,  HeadStack-separatorLine의 Spacing은 0, 이들과 TableView까지의 Spacing은 10을 줘야 Cell이 separatorLine에 딱 붙지않고 떨어지게 보였었습니다. StackView 안에서 아이템간 Spacing이 달라서 추가해줬었습니다.. 

그런데 조금 더 공부해보니 StackView 내부에서도 `setCustomSpacing(_:after:)`를 사용해서 아이템간의 spacing을 서로 다르게 줄 수 있다는 것을 알게됐습니다..!)

stevenkim18 replied:
저는 변동이 거의 없는 뷰라면 StackView보다는 UIView를 사용하는 것을 권장합니다!! 오히려 StackView를 많이 사용하면 더 복잡해 질 수 있어요!

stevenkim18 commented on ProjectManager/ProjectManager/View/MainView/MainViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        setUpNavigationBar()
+        configureHierarchy()
+        configureLayout()
+    }
```
4개의 함수들은 ViewWillAppear에서 호출 하시는 이유가 있으신가요??

sunny-maeng replied:
결론적으론 `viewDidLoad()`로 옮겼습니다! [af80fe4](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/af80fe4f81414335dfbd2c848a679fa8310a4bdd)

처음엔, View를 그리고, 계층을 만들고, Layout 하는 관련 메서드들을 분리하고 싶은게 이유였는데요
Layout Cycle이 `viewWillAppear`와 `viewDidLoad` 사이에서 일어나서 ViewWillAppear에 호출했었습니다.
그런데 다시 생각하니 `viewWillAppear`는 뷰가 불릴 때마다 호출되고, `viewDidLoad()`는 처음 한 번 호출되고 끝나니, 처음 한번만 호출해도 되는 메서드들이라 `viewDidLoad()`로 옮겨놓았습니다.

stevenkim18 replied:
꼭 `viewWillAppear` 에서만 호출을 해야되는 경우가 있을꺼에요!! 
`viewDidLoad` 시점에서 아직 뷰에 frame이 0 인 경우가 종종 있습니다!!

stevenkim18 commented on ProjectManager/ProjectManager/View/MainView/MainViewController.swift:
```diff
@@ -0,0 +1,214 @@
+final class MainViewController: UIViewController {
+    
+    private let mainViewModel = MainViewModel()
+    private let lists: [ListView] = [ListView(), ListView(), ListView()]
```
`listViews` 가 좋겠죠? ㅎㅎ

sunny-maeng replied:
View 요소들 네이밍 수정하면서 같이 수정했습니다 [67b99a3](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/67b99a3aa9dc28c5215d3119e2b9d1855ef35d82)
가독성이 한결 나아진 것 같아요 감사합니다 😁

stevenkim18 commented on ProjectManager/ProjectManager/View/MainView/ListView.swift:
```diff
@@ -0,0 +1,106 @@
+    
+    var projectList: UITableView {
+        return tableView
+    }
```
tableview임을 알 수 있는 네이밍이 좋을 것 같습니다!

sunny-maeng replied:
projectTableView로 수정했습니다 😁 [67b99a3](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/67b99a3aa9dc28c5215d3119e2b9d1855ef35d82)

stevenkim18 commented on ProjectManager/ProjectManager/View/MainView/ListCell.swift:
```diff
@@ -0,0 +1,126 @@
+    private var descriptionLabel = UILabel(font: .body, textColor: .systemGray2, numberOfLines: 3)
+    private var dateLabel = UILabel(font: .body, numberOfLines: 0)
+    private var totalView = UIView(backgroundColor: .tertiarySystemBackground,
+                                   cornerRadius: Default.radius)
```
어떤 용도의 뷰 인가요??

제가 생각하는 것이 맞다면 cell안에 있는 `contentView`로는 처리가 안되었나요??

sunny-maeng replied:
아마 생각하시는게 맞을 것 같습니다 ㅎㅎ...
stackView내부에 margin을 직접주고, ContentView와의 거리로 spacing을 줄 수 있었네요.
불필요한 View는 삭제했습니다. 감사합니다😁 [d22ccb2](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/d22ccb2664835ae7207b86db38867bde21b983c2)

stevenkim18 commented on ProjectManager/ProjectManager/ViewModel/EditingViewModel.swift:
```diff
@@ -0,0 +1,134 @@
+    
+    private var project: Project
+    private let editTargetModel: MainViewModel
+    private let isNewProject: Bool
+    private let process: Process
+    
+    private var mode: EditingMode = .editable {
+        didSet {
+            changeMode()
+        }
+    }
```
`mode`라는 변수가 있는데 `isNewProject` 라는 변수가 또 필요한가요?

sunny-maeng replied:
isNewProject 변수를 사용하지않고, mode만 사용하는것으로 수정했습니다.
isNewProject가 하던 역할은 `true`면 배열에 새로 추가하고 `false`면 기존배열에 있었던 값을 수정할 수 있게하는 `flag` 역할이었는데, 이를 없애면서 이 역할은 `MainViewModel`에서 `checkContains()`메서드를 구현해 배열에 있는지확인하고,  `save()`메서드에서 저장할 때 `checkContains()`를 활용해 기존 배열에 존재여부를 확인해 있으면 수정, 없으면 저장하도록 분기처리했습니다

stevenkim18 commented on ProjectManager/ProjectManager/ViewModel/ListCellViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+        return deadLine < today
+    }
+    
+    init(project: Project? = nil, process: Process = .todo) {
+        self.project = project
+        self.process = process
+    }
```
`project` 모델이 nil 경우도 있나요???

sunny-maeng replied:
ViewModel사용을 mainViewModel 하나로 합치면서 ListCellViewModel의 init이 제거됐습니다 ...ㅎㅎㅎ

stevenkim18 commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -0,0 +1,76 @@
+        return process.title
+    }
+    
+    private var datas: [[Project]] = [[], [], []] {
+        didSet {
+            updateDatas(datas, datasCount(datas))
+        }
+    }
```
datas도 바로 의미 와닿지 않는 네이밍 입니다.

조금 더 직관적인 이름은 없을까요??

sunny-maeng replied:
1차원 배열을 projects 라고 생각하고, 그 배열을 묶어 가지고있는 2차원 배열을 Group이라 생각해서
`projectsGroup` 으로 수정해보았습니다. 
보시기에 괜찮으신가요...!🤔 [67b99a3](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/67b99a3aa9dc28c5215d3119e2b9d1855ef35d82)

stevenkim18 replied:
data 라는 네이밍은 너무 범용적인 이름 같아서 말씀 드렸습니다 ㅎㅎ

stevenkim18 commented on ProjectManager/ProjectManager/ViewModel/MainViewModel.swift:
```diff
@@ -0,0 +1,76 @@
+        datas.map { String($0.count) }
+    }
+    
+    var newProject: Project {
+        return Project(title: Default.title,
+                       description: Default.description,
+                       date: Default.date,
+                       uuid: UUID())
+    }
+    
+    var updateDatas: ([[Project]], [String]) -> Void = { _, _ in }
+    
+    func readTitle(of process: Process) -> String {
+        return processTitles[process.index]
+    }
+    
+    func readData(in process: Process) -> [Project] {
+        return datas[process.index]
+    }
+    
+    func registerProject(_ project: Project, in process: Process) {
+        datas[process.index].append(project)
+    }
+    
+    func editProject(_ project: Project, in process: Process) {
+        datas[process.index].enumerated().forEach { index, data in
+            guard data.uuid == project.uuid else { return }
+            datas[process.index][index] = project
+        }
+    }
```
어떤 기준으로 연산 프로퍼티와 함수를 만드시나요?

제 생각에는
newProject 같은 경우 연산 프로퍼티보다 `generateNewProject`라고 해서 사용하면 더 의미가 살 것 같습니다!
어떻게 생각하시나요???

sunny-maeng replied:
사용하는 곳에서의 코드를 생각하고 연산 프로퍼티로 만들었었습니다
`let newProject: Project = mainViewModel.newProject`
`let newProject: Project = mainViewModel.generateNewProject()`
새로운 프로젝트를 만들어내니 메서드 사용이 더 좋은 것같네요 😁 [af80fe4](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/af80fe4f81414335dfbd2c848a679fa8310a4bdd)

stevenkim18 commented:
## 조언을 얻고 싶은 부분에 대한 답변
> MVVM관련
뷰모델이 다른 뷰모델을 프로퍼티로 가지고있어도 되는지 궁금합니다.
예를들면 등록된 프로젝트를 수정하려고 할 때, EditingViewModel은 MainViewModel의 어떤 데이터(Project)인지 전달받야하고, 수정이끝나면 다시 그 데이터가 이렇게 바뀌었다고 전달해줘야하는 상황이라 지금 같은경우엔,  EditingViewModel 에서 var editTargetModel: MainViewModel
을 프로퍼티로 갖고있어 이니셜라이저에서 MainViewModel을 넘겨줘야하는 로직입니다..
이렇게 구현하는게 맞는 방향인지 조언을 얻고싶습니다 .. 

뷰모델이 다른 뷰모델을 가지고 있는 것은 썩 바람직해 보이지 않습니다.
delegate 패턴을 사용해서 해결해 보면 어떨까요??

>다른 MainViewController 에게 EditViewModel을 매개변수로 넘겨줘도 되는지 궁금합니다.
롱터치액션 핸들링에서, cell에 LongPressGestureRecognizer를 추가하고 delegate로 MainViewController로 핸들링을 넘겨줄 때, 어떤 프로젝트에 대한 정보인지 넘겨주기위해 cellViewModel을 매개변수에 담아서 넘겨주는데 이 방법이 괜찮은 방법인지 조언을 얻고 싶습니다.. 

이 질문에 대해서 코드를 보면 어느 정도 이해는 가지만 정확한 질문 의도를 잘 모르겠습니다. 구두로 설명해주시면 빠르게 이해할 수 있을 것 같습니다!

sunny-maeng committed:
refactor: 프로퍼티, 타입 네이밍 수정 및 매개변수명이 메서드명과 중복되는 메서드들 네이밍수정

- datas ➡️ projectsGroup
- View아이템들 명시적인 네이밍 (ex. stack ➡️ stackView)
- Process 열거형 ➡️ ProjectState 열거형
- 매개변수명이 메서드명과 중복되는 메서드들 네이밍수정  registerProject ➡️ register / editProject  ➡️ edit

sunny-maeng committed:
refactor: 불필요한 nested View 제거

- ListView에서 요소간 spacing이 달라 추가해줬던 이중 Stack 중 한개 삭제 ➡️ 한개로 처리
- ListCell에서 margin과 radius 추가를 위해 생성했던 View제거 ➡️ ContentView와 거리로 마진 처리, stackView에서 내부 margin및 layer radius 적용

sunny-maeng committed:
refactor: 뷰 관련 메서드 호출위치 수정 및, 연산 프로퍼티(newProject)를 메서드로 수정
- 뷰관련 메서드 호출부 이동 viewWillAppear ➡️  viewDidLoad
- 연산프로퍼티 newProject ➡️ 메서드 generateNewProject()

sunny-maeng committed:
refactor: 메서드네이밍 수정(매개변수와 중복 삭제)

sunny-maeng committed:
build: GoogleService-info-plist 추가

sunny-maeng committed:
Update Step2 README

sunny-maeng committed:
refactor: 네이밍 process ➡️ state 수정으로 매개변수명 수정 및 fetchProject메서드에 index초과 시 nil반환 로직 추가

sunny-maeng committed:
Merge remote-tracking branch 'refs/remotes/origin/step2'

sunny-maeng committed:
refactor: ViewModel을 MainViewModel 하나사용, 타입간 데이터 전달 NotificationCenter 사용

sunny-maeng committed:
test: MainViewModel UnitTests 수정 및 추가 (Coverage: 100%)

sunny-maeng committed:
refactor: 클로저 축약사용 지양 (풀어서 사용)

sunny-maeng commented:
> 클로져를 많이 사용해서 코드의 가독성이 떨어진다.
> 클로져의 생략 표현들을 많이 사용하셨는데 처음 학습할 때는 풀어서 쓰는 것을 저는 권장합니다.

클로저 축약을 지양하고, 풀어쓰는 방향으로 리팩토링했습니다 [1b80b26](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/1b80b2661da2a68d48c550075f20e89f351c9fdb)

> Main뷰컨, Edit뷰컨, Cell 모두에 뷰모델이 있어서 오히려 구조가 복잡해진 느낌이었다.
Main 뷰컨에 대한 뷰모델만 있고 필요에 따라서 다른 뷰모델을 만들어도 되지 않을까..

Main뷰모델과, DTO역할을하는 Project뷰모델로 통합해보았습니다. [1177753](https://github.com/yagom-academy/ios-project-manager/pull/239/commits/11777534373358fe59495c4cae3187f0f2f11250) 

뷰모델을 통일하면서 값을 넘기는 방법을 수정했습니다
1. Editing뷰컨에서 프로젝트 수정완료후 Main뷰컨으로 넘길때:
 - 기존: Editing뷰컨 -> Editing뷰모델 -> Main뷰모델,  Main뷰컨에서는 뷰모델과 바인딩되어있는 `update클로저` 호출로 View자동업데이트
 - 현재: Editing뷰컨에서 노티피케이션 Post -> Main뷰컨에서 수신 후 Main 뷰모델 interface로 수정
 
2. LongPreesedTouch를 Cell에서 인식한 후 Main뷰컨으로 넘길때
- 기존: Cell에서 Delegate를 이용해 매개변수로 Project객체와 ProjectSate를 전달 -> Main뷰컨에서 핸들링
- 현재: Cell에서 Notification을 UserInfo에 Project객체와 ProjectSate를 담아서  Post -> Main뷰컨에서 수신 후 핸들링

sunny-maeng committed:
refactor: 네이밍수정 description ➡️ detail (CoreData Attrubute로 descripton 사용이불가로 detail로 통일하기위해서)

stevenkim18 commented:
오!! 써니쿠키!! 연휴인데도 너무 고생 많으셨습니다👍

수정 해주신 코드 잘 보았습니다.
피드백을 잘 반영해 주셔서 감사합니다🙇‍♂️

두가지 물어보고 싶어요.
1. 써니쿠키가 생각하셨을 때 첫번째 구현 방식과 두번째 구현 방식의 각각 장단점을 듣고 싶어요!
2. NotificationCenter를 사용하셨는데 이유가 궁금해요!

sunny-maeng commented:
> 첫번째 구현 방식과 두번째 구현 방식의 각각 장단점을 듣고 싶어요!

첫번째 구현방식에서는 ViewController - ViewModel 하나씩 짝 지어 있는 구현이었습니다.
- **장점:** 한 개의 View에서의 모든 로직을 한 ViewModel이 갖고있어서 한 개의 View가 수행하는 로직을 구분할 수 있다고 생각했습니다 더불어 한개의 View에 담길 Model 타입을 쉽게 파악할 수 있었습니다.
    - 예를들어 마감기한은 `var date: Date()` 타입에서 Label에 담길때는 `String`으로 바뀌는데, 뷰가 자신의 ViewModel에서 직접 `var date: String`으로 가지기 때문에, 별다른 처리없이 ViewMode의 값을 View에 바로 담을 수 있었습니다.

- **단점:** (스티븐이 말씀해 주신 것처럼) 구조가 복잡해진다는 것입니다. 
    - 모든 View에서의 핵심 로직들이 MainViewModel의 projectGroup배열을 변화시키는 로직인데, View가 본인의 ViewModel에서 핸들링해야하기 때문에, 결국 View -> 본인의 ViewModel -> MainViewModel 에게 전달에 전달을 해야하는 상황이라 복잡해진다는 것입니다.
   - 복잡한 부분은 UML에서 더 두드러지게 확인가능합니다. 

두번째 구현방식은 MainViewModel하나와, 필요에 의해 ProjectViewModel이라는 DTO 역할의 타입을 구현한 것입니다.

- **장점:** 첫번재 방법에서 단점으로 말했던 복잡해진 구조를 좀 더 단순하게 할 수 있다는 점입니다. 로직을 처리하기 위한 이벤트 전달을 확실히 짧은 경로로 명확하게 할 수 있었습니다. 코드를 읽는사람에게도 훨씬 직관적으로 다가올 것 같습니다
- **단점:** EditingViewModel에서 dateLabel의 Model로는 String타입이 필요하지만, MainViewModel에서 Date타입으로 갖고있기 떄문에 String으로 바꿔주는 처리를 View가 직접 처리해야했습니다. 추가로.. 날짜가 지나면 빨간색 처리를 하는 로직도 본인의 ViewModel이 없기 때문에 Date Extension으로 넘겼는데, 결국 필요한 로직들을 분산시킨다는 느낌이 없지않아 있었습니다.


> NotificationCenter를 사용하셨는데 이유가 궁금해요!

데이터수신자가 발신자의 정보를 모르고 이벤트를 처리하게 하려고 노티피케이션센터를 사용했습니다. 
Delegate 패턴은 다른 객체의 인스턴스를 보유하고 그 인스턴스를 활용해서(인스턴스.delegate = self) 이벤트를 처리할 수 있는데, 두 인스턴스가 의존없이 이벤트를 처리하게 하고싶었습니다.

stevenkim18 commented:
고생하셨습니다!!

아키텍쳐 패턴은 큰 틀에서 가지고 가되 정답이 없어서 써니쿠키가 직접 느낀점이 궁금해서 여쭤봤습니다(리드미에 적어도 좋을 것 같네요 ㅋㅋ)

그리고 NotificationCenter로 의존이 없게 하신 것은 매우 잘하셨습니다!
하지만 나중에 코드의 양이 많아지면 디버깅하기가 매운 단점이 있습니다!!(현업 코드에서는 사용하는 경우를 못봤어요!)

그래서 또 다른 방법들을 고민해보시면 좋겠습니다!!

머지하겠습니다!! 다음 스탭도 잘 진행해주세요!
